aptos-move/e2e-move-tests/src/tests/function_value_depth.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Tests for deeply-nested function values. The Move VM must ensure that it is not possible to
+//! construct values that are too deep, as this can cause stack overflow.
+
+use crate::{assert_success, assert_vm_status, tests::common, MoveHarness};
+use aptos_framework::BuildOptions;
+use aptos_package_builder::PackageBuilder;
+use aptos_transaction_simulation::Account;
+use aptos_types::transaction::TransactionStatus;
+use move_core_types::{account_address::AccountAddress, vm_status::StatusCode};
+
+#[test]
+fn test_vm_value_too_deep_with_function_values() {
+    let mut h = MoveHarness::new();
+    let acc = h.new_account_at(AccountAddress::from_hex_literal("0x99").unwrap());
+
+    let result = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x99::m {
+            public fun dummy1(_v: vector<vector<vector<vector<vector<vector<vector<vector<|| has drop+copy>>>>>>>>) {}
+
+            public entry fun run() {
+                let f: || has copy+drop = || {};
+
+                let i = 0;
+                while (i < 10000) {
```

Comment:
Super deep test is good for above mentioned reasons, but maybe let's also check a smaller (and above limit) that it still errors?

Reply:
Yes, still need to add unit tests and off by 1, I am pretty sure I have some depths starting at 0 and some at 1 😄🤦‍♂️

Reply:
Update: depths for visitors start at 0, but I am not sure changing to 1 is safe because we have places in `misc.rs` where we assert depth is 0 and also offset it by 1..., so I think it is fine to have 0s there - those are metering only and we should not see any overflows anyway + we propagate errors and there are no unwraps, so we should be safe.

Reply:
Added the boundary case